### PR TITLE
GenBankWriter does not use information in the original header #942

### PR DESCRIPTION
**Pull request by jamesmorris**
Hello, 
I have been working on the issue that a GenBank file read in through the parser and then written back out will result in loss of information from the original file. While looking into this I have also fixed a few other issues with the parser and writer for Genbank files. The highlights of this work includes:
- added a new writer helper method that will retain the information in the original locus line
- fixed writing out db_xref feature qualifiers which was not working
- fixed writing out the accession version and GI ID
- allow locus ids that contain spaces
- allow the locus id to be empty
- fixed writing out feature locations containing joins and partial locations
- fixed reading and writing locations on multiple lines
- Maintained quotes around feature qualifier values in files that are written out
- Added a test and test file for feature locations containing joins
- Added a test to check for differences between an original file and one that had been written out by the writer

### CODE REVIEW

File: biojava-core/src/main/java/org/biojava/nbio/core/sequence/io/GenericGenbankHeaderFormat.java [hunk 1]
```diff
@@ -222,6 +228,22 @@ private String _write_the_first_line(S sequence) {
 		 */
 	}
 
+	/**
+	 * Write the original LOCUS line.
+	 *
+	 * @param sequence
+	 */
+	private String _write_original_first_line(S sequence) {
```
Comment (josemduarte) [inline biojava-core/src/main/java/org/biojava/nbio/core/sequence/io/GenericGenbankHeaderFormat.java:236]: Nitpick: could you keep the method name as per usual camel case convention?

File: biojava-core/src/main/java/org/biojava/nbio/core/sequence/io/GenericGenbankHeaderFormat.java [hunk 2]
```diff
@@ -38,6 +39,7 @@
 	private static final int HEADER_WIDTH = 12;
 	private static final String lineSep = "%n";
 	private String seqType = null;
+	private Boolean useOriginalHeader = false;
```
Comment (josemduarte) [inline biojava-core/src/main/java/org/biojava/nbio/core/sequence/io/GenericGenbankHeaderFormat.java:42]: Any specific reason not to use the primitive `boolean`?

File: biojava-core/src/main/java/org/biojava/nbio/core/sequence/io/GenericGenbankHeaderParser.java [hunk 3]
```diff
@@ -37,15 +37,57 @@
...
 public class GenericGenbankHeaderParser<S extends AbstractSequence<C>, C extends Compound> implements SequenceHeaderParserInterface<S,C> {
 
+	@SuppressWarnings("unused")
+
+	// Brief description of sequence; includes information such as source organism,
+	// gene name/protein name, or some description of the sequence's function (if
+	// the sequence is non-coding). If the sequence has a coding region (CDS),
+	// description may be followed by a completeness qualifier, such as "complete
+	// cds".
+	private String description;
```
Comment (josemduarte) [inline biojava-core/src/main/java/org/biojava/nbio/core/sequence/io/GenericGenbankHeaderParser.java:47]: Nice docs! Could you make them into javadocs? i.e. surround comments by `/**   */` . The same for others below

**Review by josemduarte — CHANGES_REQUESTED** [review]
Very nice, thank you! A few minor comments

File: biojava-core/src/main/java/org/biojava/nbio/core/sequence/io/GenericGenbankHeaderFormat.java (hunk 1 above)
Comment (jamesmorris) [reply biojava-core/src/main/java/org/biojava/nbio/core/sequence/io/GenericGenbankHeaderFormat.java:236]: Sure happy to change, I would normally but I was just following the convention already used in the class
e.g. _write_single_line, _write_multi_line etc

File: biojava-core/src/main/java/org/biojava/nbio/core/sequence/io/GenericGenbankHeaderFormat.java (hunk 2 above)
Comment (jamesmorris) [reply biojava-core/src/main/java/org/biojava/nbio/core/sequence/io/GenericGenbankHeaderFormat.java]: No, I have now fixed this

File: biojava-core/src/main/java/org/biojava/nbio/core/sequence/io/GenericGenbankHeaderParser.java (hunk 3 above)
Comment (jamesmorris) [reply biojava-core/src/main/java/org/biojava/nbio/core/sequence/io/GenericGenbankHeaderParser.java:49]: Sure this is now done

File: biojava-core/src/main/java/org/biojava/nbio/core/sequence/io/GenericGenbankHeaderFormat.java (hunk 1 above)
Comment (josemduarte) [reply biojava-core/src/main/java/org/biojava/nbio/core/sequence/io/GenericGenbankHeaderFormat.java:236]: Oh I didn't realise that was the style already for that file. Well I guess we can live with that a bit longer...

**Comment by emckee2006** [comment]
It was initially ported from biopython in a very rudimentary fashion, so
the function names pretty much match the equivalent biopython ones....

**Review by josemduarte — APPROVED** [review]
Thanks! LGTM

**Pull request merged**